src/bot/SoundBot.ts

```diff
@@ -43,53 +43,55 @@ export default class SoundBot extends Client {
   private addEventListeners() {
     this.on('ready', this.onReady);
     this.on('message', this.onMessage);
-    this.on('voiceStateUpdate', this.onUserLeavesVoiceChannel);
-    this.on('voiceStateUpdate', this.onUserJoinsVoiceChannel);
+    this.on('voiceStateUpdate', this.handleVoiceStateUpdate);
     this.on('guildCreate', this.onBotJoinsServer);
   }
 
   private onReady() {
-    if (!this.user) {
-      throw new Error('User not set after ready.');
-    }
-    this.user.setActivity(this.config.game);
-    this.commands.registerUserCommands(this.user);
+    this.user!.setActivity(this.config.game);
+    this.commands.registerUserCommands(this.user!);
   }
 
-  private onUserJoinsVoiceChannel(prevState: GuildMember, user: GuildMember) {
-    if (!this.user || user.id === this.user.id) return;
+  private handleVoiceStateUpdate(oldState: VoiceState, newState: VoiceState) {
+    if (!oldState.channelID && newState.channelID) {
+      return this.onUserJoinsVoiceChannel(newState.channel!, newState.member!.user);
+    }
 
-    if (!user.voice) return;
-    if (!entrances.exists(user.id)) return;
+    if (oldState.channelID && !newState.channelID) {
+      return this.onUserLeavesVoiceChannel(oldState.channel!, oldState.member!.user);
+    }
+  }
+
+  private onUserJoinsVoiceChannel(channel: VoiceChannel, user: User) {
+    if (!entrances.exists(user.id)) {
```

Comment:
Please keep the short one line returns if it is just `return;` and nothing else.